File: schemas/marketplace/mongodb/balance_transactions.js
// balance_transactions: Ledger of all balance-affecting transactions per vendor.
// See README.md for full design rationale.

const mongoose = require("mongoose");

const balanceTransactionSchema = new mongoose.Schema(
  {
    vendor_id: { type: mongoose.Schema.Types.ObjectId, ref: "Vendor", required: true },
    type: {
      type: String,
      enum: ["earning", "commission", "payout", "refund", "adjustment", "hold", "release"],
      required: true,
    },
    amount: { type: Number, required: true },
    currency: { type: String, required: true },
    running_balance: { type: Number, required: true },
    reference_type: { type: String, default: null },
    reference_id: { type: String, default: null },
    description: { type: String, default: null },
  },
  { timestamps: { createdAt: "created_at", updatedAt: false } }
);

balanceTransactionSchema.index({ vendor_id: 1, created_at: 1 });
balanceTransactionSchema.index({ type: 1 });
balanceTransactionSchema.index({ reference_type: 1, reference_id: 1 });

module.exports = mongoose.model("BalanceTransaction", balanceTransactionSchema);

File: schemas/marketplace/spacetimedb/balance_transactions.rs
// balance_transactions: Ledger of all financial movements affecting vendor balances.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum BalanceTransactionType {
    Earning,
    Commission,
    Payout,
    Refund,
    Adjustment,
    Hold,
    Release,
}

// Composite index: (vendor_id, created_at)

#[spacetimedb::table(name = balance_transactions, public)]
pub struct BalanceTransaction {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub vendor_id: String, // UUID, FK -> vendors.id (restrict delete)

    pub transaction_type: BalanceTransactionType,
    pub amount: i32,
    pub currency: String,
    pub running_balance: i32,
    pub reference_type: Option<String>,
    pub reference_id: Option<String>,
    pub description: Option<String>,
    pub created_at: Timestamp,
}
